README.md
# BidFeed

[![Build Status](https://travis-ci.org/tpg/bidfeed.svg?branch=master)](https://travis-ci.org/tpg/bidfeed)

__Bid or Buy XML Feed Library__

BidFeed is an XML generator for the Bid or Buy XML feed.

## Installation

BidFeed should be installed through Composer:

```bash
composer require thepublicgood/bidfeed
```

## Usage
You'll need to `require` the Composer autoloader if you're not using a framework that does so automatically.

Start by creating a new instance of `TPG\BidFeed\Builder`. The methods on the `Builder`.

```php
require __DIR__.'/vendor/autoload.php';

$feed = new TPG\BidFeed\Builder();
```

The `Builder` class provides a `products()` method that returns a `TPG\BidFeed\Collection` instance. You can add new products to the feed by pushing new instances of `TPG\BidFeed\Product`.

```php
$product = new Product();

$feed->products()->push($product);
```

## Generating the XML
At any point, you can get a copy of the XML by calling the `toXml()` method on the `Builder` instance.

```php
$xml = $feed->toXml();

// You can also save it directly to a file by passing a filename:
$feed->toXml('feed.xml');
```

Calling the `toXml()` method does not alter the instance in any way, so you can continue to build up products even after generating an XML output.

The `toXml()` method will also call the `verifyAttributes()` method on each `Product` instance. If there is anything missing, a `MissingRequiredAttribute` exception will be thrown.

## Adding Products
You build up a feed by adding products. The `Product` class provides a simple API for working with products.

A single product is represented by an instance of `Product`. the `Builder::products()` method always returns a `Collection` instance containing a set of `Product` instances.

```php
// You can...
$feed->products()->push((new Product())->...);

// or add multiple products at once...
$feed->products()->push([$product1, $product2]);
```

## Required Attributes

```php
$product->name($productName)
    ->code($productCode)
    ->category($category)
    ->price($price, $marketPrice)
    ->availableQuantity($quantity)
    ->description($description);
```

Some attributes are generally not required unless the product needs to appear in Google Ads posted by Bid or Buy, then these attributes MUST be supplied:

```php
$product->gtin($productGtin)        // GTIN (GTIN-12 or GTIN-13)
    ->mpn($mpn);                    // MPN
```

A number of attributes are required. All products must have a name, a product code, a category, price, quantity, and description. If any of these attributes are missing on a single product, a `MissingRequiredAttribute` exception will be thrown.

## Setting a product code

A product MUST have a unique product code and it cannot change. Any duplicate product codes will be ignored an the product will not be imported. The product code has a maximum length of 100 characters.

```php
$product->code('CODE-000111');
```

## Setting a product name
All products must be named. You can specify the product name using the `name` method. Product names cannot be longer than 100 characters:

```php
$product->name('My Product');
```

## Setting a description
The description of the product should include the details of the product. It cano be upto 8000 characters long and can include some limited HTML. The BidFeed library currently allows `P` and `BR` html elements.

```php
$product->description('<p>Product description</p>');
```

## Setting a category
A product category must be supplied with app products and preferably based on Google's taxonomy. See [https://support.google.com/merchants/answer/6324436](). It's recommended that the full category path be specified. The `category` method accepts an array of category names. So if the category should be `Electroncis - Laptops - Apple` you can do:

```php
$product->category(['Electronics', 'Laptops', 'Apple']);
```

## Setting Price
All products MUST have a price set. There are two prices that a product can have. The actual selling price, which is required, and a market price, which is optional. The market price is only for reference and not used as a selling price in any way.

You can set both prices by passing the selling price, and the market price at floats to the `price()` method, or set the market price separately using the `marketPrice()` method.

```php
// Set a selling price
$product->price(199.95);

// Set a selling price and market price
$product->price(199.95, 219.95);

// Set a marketing price
$product->marketPrice(211.50);
```

## Available Quantity
Bid or Buy require that an available quantity be specified. You can specify the quantities using the `availableQuantity()` method. Any product with a quantity of 0 will not be imported.

```php
$product->availableQuantity(10);
```

## Product Images
Images must included with all products. At least one image must be provided. You can provide more than one, but only the first image will be used as the product cover image. Images are displayed by Bid or Buy in the order they are provided.

You can provide a set of image URLs by passing an array to the `images()` method.

```php
$product->images([$image1, $image2]);
```

Images are stored as a `Collection` instance. You can get to the instance using the `imageCollection()` method.

```php
// Add an image to the collection
$product->imageCollection()->add($imageUrl3);
```

## Product Attributes
Products can have an optional set of custom attributes. It's common to add a `Brand` attribute here, but you can add any attributes needed. According to the Bid or Buy feed spec, attributes that do not match marketplace values are ignored, so you should only add attributes that are category specific.

```php
$product->productAttributes([
    'Brand' => 'Apple'
]);
```

## Guarantees and Warranties
Products can have a gaurantee and/or a warranty set. You can set gaurantees using the `guarantee` method, and warranties can be set using the `warranty` method. The signature for both methods is identical:

```php
$product->guarantee($type, $text);
$product->warranty($type, $text);
```

BidFeed provides a set of constants for gaurantee and warranty types:

```php
GUARANTEE_NOT_OFFERED;      // No guarantee
GUARANTEE_MONEY_BACK_7;     // 7 day money back guarantee
GUARANTEE_MONEY_BACK_10;    // 10 day
GAURANTEE_MONEY_BACK_15;    // 15 day
GUARANTEE_MONEY_BACK_30;    // 30 day
GUARANTEE_REPLACEMENT_7;    // 7 day replacement guarantee
GUARANTEE_REPLACEMENT_10;   // 10 day
GAURANTEE_REPLACEMENT_15;   // 15 day
GUARANTEE_REPLACEMENT_30;   // 30 day

WARRANTY_NOT_OFFERED;       // No warranty
WARRANTY_REPLACEMENT;       // Replacement warranty
WARRANTY_DEALER;            // Dealers warranty
WARRANTY_MANUFACTURER;      // Manufacturers warranty
```

The second parameter allows you to specify more detail around the guarantee or warranty. The text MUST be provided if the first parameter is anything but `GUARANTEE_


## Shipping Class


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.3] - 2019-06-26
### Fixed
- Fixed a bug that was not setting the available quantity using the wrong node.

## [0.0.2] - 2019-06-24
### Fixed
- Fixed a bug in the `Builder->toXml()` that caused an error when no collection had been added.

## [0.0.1] - 2019-06-20
- Add products to an XML feed builder.
- Render the feed as an array.
- Render the feed as XML.


tests/TestCase.php
<?php

namespace TPG\Test;

class TestCase extends \PHPUnit\Framework\TestCase
{
}


tests/BuilderTest.php
<?php

namespace TPG\Test;

use DateTime;
use TPG\BidFeed\Builder;
use TPG\BidFeed\Product;
use TPG\BidFeed\Collection;

class BuilderTest extends TestCase
{
    /**
     * @param int $count
     * @return Product[]
     */
    private function createProduct($count = 1)
    {
        $products = [];
        for ($i = 0; $i < $count; $i++) {
            $products[] = (new Product())
                ->code($code = rand(1000, 2000))
                ->name('Test Product '.$code)
                ->category(['Electronics', 'Laptops', 'Apple'])
                ->price(199.95)
                ->availableQuantity(10)
                ->images([
                    'https://test.test/images/image1.jpg',
                ])
                ->description('The product description');
        }

        return $products;
    }

    /**
     * @test
     */
    public function it_can_have_version_information()
    {
        $feed = new Builder();

        $now = new DateTime();
        $feed->version('BidFeed v1.0.0');
        $feed->created($now);

        $this->assertEquals('BidFeed v1.0.0', $feed->toArray()['Version']);
        $this->assertEquals($now->format(DATE_ATOM), $feed->toArray()['ExportCreated']);
    }

    /**
     * @test
     */
    public function it_can_create_a_new_feed()
    {
        $feed = new Builder();

        $feed->products()->push($product = $this->createProduct()[0]);

        $this->assertCount(1, $feed->products());
        $this->assertEquals($product->toArray()['ProductName'], $feed->products()->first()->toArray()['ProductName']);
    }

    /**
     * @test
     */
    public function a_product_can_have_an_image_collection()
    {
        $product = $this->createProduct()[0];

        $this->assertInstanceOf(Collection::class, $product->imageCollection());
        $this->assertCount(1, $product->imageCollection());
    }

    /**
     * @test
     */
    public function collection_items_have_the_collection_as_a_parent()
    {
        $builder = new Builder();

        $builder->products()->push($this->createProduct());

        $this->assertInstanceOf(Collection::class, $builder->products()->first()->parent());
    }

    /**
     * @test
     */
    public function a_product_can_be_deleted_from_the_collection()
    {
        $builder = new Builder();

        $builder->products()->push($this->createProduct());

        $builder->products()->first()->delete();

        $this->assertCount(0, $builder->products());
    }

    /**
     * @test
     */
    public function a_product_can_have_custom_attributes()
    {
        $builder = new Builder();

        $product = $this->createProduct()[0];
        $product->productAttributes([
            'Brand' => 'Products Inc.',
        ]);

        $builder->products()->push($product);

        $this->assertEquals('Products Inc.', $builder->products()->first()->productAttributesCollection()->find('Brand'));
    }

    /**
     * @test
     */
    public function the_builder_can_generate_xml()
    {
        $builder = new Builder();

        $products = $this->createProduct(2);
        $products[0]->productAttributes([
            'Brand' => 'Products Inc.',
        ]);

        $builder->products()->push($products);

        $xml = simplexml_load_string($builder->toXml(__DIR__.'/test.xml'));

        $this->assertEquals($products[0]->toArray()['ProductName'], (string) $xml->Products->Product[0]->ProductName);
        $this->assertEquals($products[1]->toArray()['ProductCode'], (string) $xml->Products->Product[1]->ProductCode);

        $this->assertEquals('Products Inc.', $xml->Products->Product[0]->ProductAttributes->Brand);

        $this->assertFileExists(__DIR__.'/test.xml');
        unlink(__DIR__.'/test.xml');
    }

    /**
     * @test
     */
    public function products_can_have_guarantees_and_warranties()
    {
        $product = $this->createProduct()[0];
        $product->guarantee(Product::GUARANTEE_MONEY_BACK_10, '10 day money back guarantee');
        $product->warranty(Product::WARRANTY_DEALER, 'Dealer warranty');

        $builder = new Builder();
        $builder->products()->push($product);

        $this->assertEquals(10, $builder->products()->first()->toArray()['GuaranteeType']);
        $this->assertEquals('10 day money back guarantee', $builder->products()->first()->toArray()['GuaranteeText']);
        $this->assertEquals(2, $builder->products()->first()->toArray()['WarrantyType']);
        $this->assertEquals('Dealer warranty', $builder->products()->first()->toArray()['WarrantyText']);
    }
}


src/Builder.php
<?php

namespace TPG\BidFeed;

use DateTime;

class Builder
{
    /**
     * @var Collection
     */
    protected $products;

    /**
     * @var string
     */
    protected $version = 'BidFeed Library v0.1.0';

    /**
     * @var DateTime
     */
    protected $created;

    /**
     * Builder constructor.
     */
    public function __construct()
    {
        $this->created = new DateTime();
    }

    /**
     * Set the "plugin" version number.
     *
     * @param string $version
     * @return Builder
     */
    public function version(string $version): self
    {
        $this->version = $version;

        return $this;
    }

    /**
     * Set the time the feed was created.
     *
     * @param DateTime $created
     * @return Builder
     */
    public function created(DateTime $created): self
    {
        $this->created = $created;

        return $this;
    }

    public function products(): Collection
    {
        return $this->products ?: $this->products = new Collection();
    }

    public function toArray(): array
    {
        return [
            'Version' => $this->version,
            'ExportCreated' => $this->created->format(DATE_ATOM),
            'Products' => $this->products()->toArray(),
        ];
    }

    public function toJson($pretty = false): string
    {
        return json_encode($this->toArray(), JSON_NUMERIC_CHECK + ($pretty ? JSON_PRETTY_PRINT : 0));
    }

    public function toXml(string $filename = null)
    {
        $document = new \DOMDocument();

        $document->encoding = 'UTF-8';

        $root = $document->createElement('ROOT');

        $document->appendChild($root);

        $this->addVersionNode($root);
        $this->addCreatedNode($root);

        $this->products()->toXml($root);

        if ($filename) {
            $document->save($filename);
        }

        return $document->saveXML();
    }

    protected function addVersionNode(\DOMElement $root)
    {
        $versionNode = new \DOMElement('Version');
        $root->appendChild($versionNode);
        $versionNode->textContent = $this->version;
    }

    protected function addCreatedNode(\DOMElement $root)
    {
        $createdNode = new \DOMElement('ExportCreated');
        $root->appendChild($createdNode);
        $createdNode->textContent = $this->created->format(DATE_ATOM);
    }
}


src/Collection.php
<?php

namespace TPG\BidFeed;

use Countable;
use ArrayAccess;
use TPG\BidFeed\Traits\Arrayable;
use TPG\BidFeed\Traits\Iteratable;

class Collection implements ArrayAccess, Countable, \Iterator
{
    use Arrayable, Iteratable;

    /**
     * @var array
     */
    protected $items;

    /**
     * Collection constructor.
     * @param ArrayAccess|array|null $items
     */
    public function __construct($items = null)
    {
        $this->items = $items ?: [];
    }

    /**
     * Count the collection items.
     *
     * @return int
     */
    public function count()
    {
        return count($this->items);
    }

    /**
     * Check if a key exists in the collection.
     *
     * @param $key
     * @return bool
     */
    public function has($key): bool
    {
        return $this->offsetExists($key);
    }

    /**
     * Get the first item in the collection.
     *
     * @return mixed|null
     */
    public function first()
    {
        return $this->count() ?
            $this->items[array_keys($this->items)[0]] :
            null;
    }

    /**
     * Get the last item in the collection.
     *
     * @return mixed|null
     */
    public function last()
    {
        return $this->count() ?
            $this->items[array_reverse(array_keys($this->items))[0]] :
            null;
    }

    /**
     * Find an item by its key.
     *
     * @param mixed $key
     * @return mixed|null
     */
    public function find($key)
    {
        return $this->has($key) ? $this->items[$key] : null;
    }

    /**
     * Push an item onto the collection.
     *
     * @param $item
     * @return $this
     */
    public function push($item)
    {
        if (! is_array($item)) {
            $item = [$item];
        }
        $this->items = array_merge($this->items, $item);

        $this->updateParent();

        return $this;
    }

    /**
     * Insert an item at the top of the collection.
     *
     * @param $item
     * @return $this
     */
    public function unshift($item)
    {
        array_unshift($this->items, $item);

        return $this;
    }

    /**
     * Pop the last item off the collection.
     *
     * @return mixed
     */
    public function pop()
    {
        return array_pop($this->items);
    }

    /**
     * Shift the first item off the top of the collection.
     *
     * @return mixed
     */
    public function shift()
    {
        return array_shift($this->items);
    }

    /**
     * Delete an item from the collection.
     *
     * @param $item
     */
    public function delete($item)
    {
        $index = array_search($item, $this->items);

        if ($index !== false) {
            $this->offsetUnset($index);
        }
    }

    /**
     * Loop other each item in the collection.
     *
     * @param callable $callback
     */
    public function each(callable $callback)
    {
        foreach ($this->items as $item) {
            $callback($item);
        }
    }

    /**
     * Loop over each item in the collection and apply the result of the callback.
     *
     * @param callable $callback
     * @return Collection
     */
    public function map(callable $callback): self
    {
        $items = [];
        foreach ($this->items as $item) {
            $items[] = $callback($item);
        }

        return new self($items);
    }

    /**
     * Filter the items in the collection by the result of the callback.
     *
     * @param callable $callback
     * @return Collection
     */
    public function filter(callable $callback): self
    {
        $items = [];
        foreach ($this->items as $item) {
            if ($callback($item)) {
                $items[] = $item;
            }
        }

        return new self($items);
    }

    /**
     * Return array of collected items.
     *
     * @return array
     */
    public function toArray(): array
    {
        return [
            'Products' => array_map(function (Product $product) {
                return $product->toArray();
            }, $this->items),
        ];
    }

    public function toJson($pretty = false): string
    {
        return json_encode($this->toArray(), JSON_NUMERIC_CHECK + ($pretty ? JSON_PRETTY_PRINT : 0));
    }

    public function toXml(\DOMElement $root)
    {
        $element = $root->appendChild(new \DOMElement('Products'));
        foreach ($this->items as $product) {
            $productElement = $element->appendChild(new \DOMElement('Product'));
            $root->appendChild($element);
            $product->toXmlNode($productElement);
        }

        return $root;
    }

    /**
     * Update the parent on collected items.
     */
    protected function updateParent()
    {
        $this->each(function ($item) {
            if (method_exists($item, 'parent')) {
                $item->parent($this);
            }
        });
    }
}


src/Traits/Arrayable.php
<?php

namespace TPG\BidFeed\Traits;

trait Arrayable
{
    /**
     * Check if an item exists.
     *
     * @param mixed $offset
     * @return bool
     */
    public function offsetExists($offset)
    {
        return isset($this->items[$offset]);
    }

    /**
     * Get an item.
     *
     * @param mixed $offset
     * @return mixed
     */
    public function offsetGet($offset)
    {
        return $this->items[$offset];
    }

    /**
     * Get an item.
     *
     * @param mixed $offset
     * @param mixed $value
     */
    public function offsetSet($offset, $value)
    {
        $this->items[$offset] = $value;

        if (method_exists($value, 'parent')) {
            $value->parent($this);
        }
    }

    /**
     * Remove an item.
     *
     * @param mixed $offset
     */
    public function offsetUnset($offset)
    {
        unset($this->items[$offset]);
    }
}


src/Traits/Iteratable.php
<?php

namespace TPG\BidFeed\Traits;

trait Iteratable
{
    /**
     * Get the current item.
     *
     * @return mixed
     */
    public function current()
    {
        return current($this->items);
    }

    /**
     * Move to the next item.
     *
     * @return mixed
     */
    public function next()
    {
        return next($this->items);
    }

    /**
     * Return the key of the current item.
     *
     * @return int|string|null
     */
    public function key()
    {
        return key($this->items);
    }

    /**
     * Check if the current item position is valid.
     *
     * @return bool
     */
    public function valid()
    {
        $key = key($this->items);

        return $key !== null && $key !== false;
    }

    /**
     * Reset the iterator to the start position.
     */
    public function rewind()
    {
        reset($this->items);
    }
}


src/Traits/HasAttributes.php
<?php

namespace TPG\BidFeed\Traits;

use TPG\BidFeed\Exceptions\MissingRequiredAttribute;

trait HasAttributes
{
    /**
     * @var array
     */
    protected $attributes = [];

    /**
     * @var array
     */
    protected $requiredAttributes = [];

    /**
     * @var null|string
     */
    protected $rootKey = null;

    /**
     * Verify the current attributes.
     *
     * @return bool
     * @throws MissingRequiredAttribute
     */
    public function verifyAttributes()
    {
        foreach ($this->requiredAttributes as $required) {
            if (! in_array($required, array_keys($this->attributes))) {
                throw new MissingRequiredAttribute($required);
            }
        }

        return true;
    }

    /**
     * Output an array.
     *
     * @return array
     */
    public function toArray(): array
    {
        return $this->attributes;
    }
}


src/Product.php
<?php

namespace TPG\BidFeed;

use TPG\BidFeed\Traits\Collectable;
use TPG\BidFeed\Traits\HasAttributes;
use TPG\BidFeed\Exceptions\MissingRequiredAttribute;

class Product
{
    use HasAttributes, Collectable;

    const CONDITION_NEW = 'New';
    const CONDITION_SECONDHAND = 'Secondhand';
    const CONDITION_REFURBISHED = 'Refurbished';

    const GUARANTEE_NOT_OFFERED = 0;
    const GUARANTEE_MONEY_BACK_7 = 7;
    const GUARANTEE_MONEY_BACK_10 = 10;
    const GUARANTEE_MONEY_BACK_15 = 15;
    const GUARANTEE_MONEY_BACK_30 = 30;
    const GUARANTEE_REPLACEMENT_7 = 1007;
    const GUARANTEE_REPLACEMENT_10 = 1010;
    const GUARANTEE_REPLACEMENT_15 = 1015;
    const GUARANTEE_REPLACEMENT_30 = 1030;

    const WARRANTY_NOT_OFFERED = 0;
    const WARRANTY_REPLACEMENT = 1;
    const WARRANTY_DEALER = 2;
    const WARRANTY_MANUFACTURER = 3;

    /**
     * @var Collection
     */
    protected $images;

    /**
     * @var Collection
     */
    protected $productAttributes;

    /**
     * @var array
     */
    protected $textFields = [
        'ProductName',
        'Category',
        'ProductMPN',
        'Description',
    ];

    /**
     * Product constructor.
     */
    public function __construct()
    {
        $this->requiredAttributes = [
            'Category',
            'Price',
            'ProductName',
            'ProductCode',
            'AvailableQty',
        ];
    }

    /**
     * Set the product code.
     *
     * @param string $code
     * @return Product
     */
    public function code(string $code): self
    {
        $this->attributes['ProductCode'] = substr(strip_tags($code), 0, 100);

        return $this;
    }

    /**
     * Set the product GTIN.
     *
     * @param string $gtin
     * @return Product
     */
    public function gtin(string $gtin): self
    {
        $this->attributes['ProductGTIN'] = strip_tags($gtin);

        return $this;
    }

    /**
     * Set the product MPN.
     *
     * @param string $mpn
     * @return Product
     */
    public function mpn(string $mpn): self
    {
        $this->attributes['ProductMPN'] = strip_tags($mpn);

        return $this;
    }

    /**
     * Set a product name.
     *
     * @param string $name
     * @return Product
     */
    public function name(string $name): self
    {
        $this->attributes['ProductName'] = $name;

        return $this;
    }

    /**
     * Set a category.
     *
     * @param array $signature
     * @return Product
     */
    public function category(array $signature): self
    {
        $this->attributes['Category'] = implode(' - ', $signature);

        return $this;
    }

    /**
     * Set a selling price.
     *
     * @param float $price
     * @param float|null $marketPrice
     * @return Product
     */
    public function price(float $price, float $marketPrice = null): self
    {
        $this->attributes['Price'] = $price;

        if ($marketPrice) {
            $this->marketPrice($marketPrice);
        }

        return $this;
    }

    /**
     * Set a market price.
     *
     * @param float $marketPrice
     * @return Product
     */
    public function marketPrice(float $marketPrice): self
    {
        $this->attributes['MarketPrice'] = $marketPrice;

        return $this;
    }

    /**
     * Set if offers should be allowed.
     * @param bool $allow
     * @return Product
     */
    public function allowOffers($allow = true): self
    {
        $this->attributes['AllowOffer'] = $allow ? 'true' : 'false';

        return $this;
    }

    /**
     * Set the available quantity.
     *
     * @param int $quantity
     * @return Product
     */
    public function availableQuantity(int $quantity): self
    {
        $this->attributes['AvailableQty'] = $quantity;

        return $this;
    }

    /**
     * Set the product description.
     *
     * @param string $condition
     * @return Product
     */
    public function condition(string $condition): self
    {
        $this->attributes['Condition'] = $condition;

        return $this;
    }

    /**
     * Set a product description.
     *
     * @param string $description
     * @return Product
     */
    public function description(string $description): self
    {
        $this->attributes['Description'] = substr(strip_tags($description, 'p,br'), 0, 8000);

        return $this;
    }

    /**
     * Add images to the Product.
     *
     * @param array $images
     * @return Product
     */
    public function images(array $images): self
    {
        $this->images = new Collection($images);

        return $this;
    }

    /**
     * Get the images Collection instance.
     *
     * @return Collection
     */
    public function imageCollection(): Collection
    {
        return $this->images ?: $this->images = new Collection();
    }

    /**
     * Add optional product attributes.
     *
     * @param array $attributes
     * @return Product
     */
    public function productAttributes(array $attributes): self
    {
        $this->productAttributes = new Collection($attributes);

        return $this;
    }

    /**
     * @return Collection
     */
    public function productAttributesCollection(): Collection
    {
        return $this->productAttributes ?: $this->productAttributes = new Collection();
    }

    /**
     * Set the product guarantee type and description.
     *
     * @param int $type
     * @param string $text
     * @return $this
     */
    public function guarantee(int $type, string $text)
    {
        $this->attributes['GuaranteeType'] = $type;
        $this->attributes['GuaranteeText'] = substr($text, 0, 300);

        return $this;
    }

    /**
     * Set the product warranty type and description.
     *
     * @param int $type
     * @param string $text
     * @return $this
     */
    public function warranty(int $type, string $text)
    {
        $this->attributes['WarrantyType'] = $type;
        $this->attributes['WarrantyText'] = substr($text, 0, 300);

        return $this;
    }

    /**
     * Set the product shipping class.
     *
     * @param string $class
     * @return $this
     */
    public function shippingProductClass(string $class)
    {
        $this->attributes['ShippingProductClass'] = $class;

        return $this;
    }

    /**
     * Set the marketplace location.
     *
     * @param string $location
     * @return $this
     */
    public function location(string $location)
    {
        $this->attributes['Location'] = $location;

        return $this;
    }

    /**
     * Verify the images.
     *
     * @throws MissingRequiredAttribute
     */
    public function verifyImages()
    {
        if ($this->images->count() === 0) {
            throw new MissingRequiredAttribute('ImageURL');
        }
    }

    /**
     * Return the product as an array.
     *
     * @return array
     */
    public function toArray(): array
    {
        return array_filter($this->attributes, function ($attributes) {
            return $attributes ? true : false;
        });
    }

    /**
     * Add attributes to the XML element.
     *
     * @param \DOMNode $node
     * @param array|null $attributes
     */
    protected function addAttributesToXmlElement(\DOMNode $node, array $attributes = null)
    {
        foreach ($attributes as $key => $value) {
            if ($value) {
                $productNode = new \DOMElement($key);

                $node->appendChild($productNode);

                if (is_array($value)) {
                    $this->addAttributesToXmlElement($productNode, $value);
                } else {
                    if (in_array($key, $this->textFields)) {
                        $value = $node->ownerDocument->createCDATASection($value);
                        $productNode->appendChild($value);
                    } else {
                        $productNode->textContent = $value;
                    }

                    $node->appendChild($productNode);
                }
            }
        }
    }

    /**
     * Add images to the XML element.
     *
     * @param \DOMNode $node
     */
    protected function addImagesToXmlElement(\DOMNode $node)
    {
        $imagesNode = new \DOMElement('Images');
        $node->appendChild($imagesNode);

        foreach ($this->images as $image) {
            $imageNode = new \DOMElement('ImageURL');
            $imagesNode->appendChild($imageNode);
            $value = $node->ownerDocument->createCDATASection($image);
            $imageNode->appendChild($value);
        }
    }

    /**
     * Add the product attributes to the XML element.
     *
     * @param \DOMNode $node
     */
    public function addProductAttributesToXmlElement(\DOMNode $node)
    {
        $attributesNode = new \DOMElement('ProductAttributes');
        $node->appendChild($attributesNode);

        foreach ($this->productAttributes as $key => $value) {
            $attribute = new \DOMElement($key);
            $attributesNode->appendChild($attribute);
            $attribute->textContent = $value;
        }
    }

    /**
     * @param \DOMNode $node
     * @return void
     * @throws MissingRequiredAttribute
     */
    public function toXmlNode(\DOMNode $node)
    {
        $this->verifyAttributes();
        $this->verifyImages();
        $this->addAttributesToXmlElement($node, $this->attributes);
        $this->addImagesToXmlElement($node);

        if ($this->productAttributes && $this->productAttributes->count()) {
            $this->addProductAttributesToXmlElement($node);
        }
    }
}
